lib/self-evict.js

```diff
@@ -0,0 +1,303 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+var _ = require('underscore');
+var TypedError = require('error/typed');
+
+var errors = require('./errors');
+var Member = require('./membership/member');
+
+
+/**
+ * @typedef {Object} Phase
+ * @property {SelfEvict.PhaseNames} phase the phase
+ * @property {number} ts the timestamp when the phase started
+ * @property {number} [duration] the duration of a phase
+ */
+
+/**
+ * A pre or post self eviction hook.
+ * @callback SelfEvict~Hook
+ * @param {SelfEvict~HookCallback} cb the callback to be called when the hook completes
+ */
+
+/**
+ * A callback used when calling Hooks
+ * @callback SelfEvict~HookCallback
+ */
+
+/**
+ * Self eviction provides a way to graceful shut down a ringpop service.
+ * A self eviction  consists of the following phases:
+ * 1. PreEvict: the preEvict hooks are invoked;
+ * 2. Evict: The node will mark  itself as faulty causing an eviction out of the membership;
+ * 3. PostEvict: the postEvict hooks are invoke;
+ * 4. Done: self eviction is done and ringpop is ready to be shut down.
+ *
+ * @param {RingPop} ringpop
+ * @return {SelfEvict}
+ * @constructor
+ */
+function SelfEvict(ringpop) {
+    if (!(this instanceof SelfEvict)) {
+        return new SelfEvict(ringpop);
+    }
+
+    /**
+     * The ringpop instance
+     * @type {RingPop}
+     * @private
+     */
+    this.ringpop = ringpop;
+
+    /**
+     *
+     * @type {Object.<string, SelfEvict~Hook>}
+     */
+    this.preEvictHooks = {};
+    this.postEvictHooks = {};
+
+    /**
+     * The callback to be called after self eviction is complete.
+     * @type {SelfEvict~callback}
+     * @private
+     */
+    this.callback = null;
+
+    /**
+     * The history of phases
+     * @type {[Phase]}
+     */
+    this.phases = [];
+}
+
+/**
+ * Get the current  phase or null when self eviction isn't initiated
+ * @return {?Phase}
+ */
+SelfEvict.prototype.currentPhase = function currentPhase() {
+    if (this.phases && this.phases.length > 0) {
```

Comment:
Should we be that defensive here? `this.phases` is initialized in the constructor to be an array...